core/src/Cuda/Kokkos_Cuda_Half.hpp

```diff
@@ -212,6 +225,70 @@ class half_t {
   KOKKOS_FUNCTION
   half_t(unsigned long long rhs) : val(cast_to_half(rhs).val) {}
 
+  // Atomics
+  KOKKOS_FUNCTION
+  void atomic_add(const volatile half_t src) const volatile {
+#ifdef __CUDA_ARCH__
+#if (CUDA_VERSION < 10000)
+    alignas(4) float lhs_tmp = __half2float(val),
+                     rhs_tmp = __half2float(src.val);
+    volatile float* lhs      = &lhs_tmp;
+    volatile float* rhs      = &rhs_tmp;
+    val                      = __float2half(lhs[0] + rhs[0]);
+#else
+    // Cuda 10 supports __half volatile stores but not volatile arithmetic
+    // operands so we use atomicAdd for cuda 10.0 instead.
+    atomicAdd(const_cast<impl_type*>(&val),
+              const_cast<const impl_type&>(src.val));
+#endif  // CUDA_VERSION
+#else
+    // 32bit arithmetic on host followed by 16bit atomic store.
+    uint16_t* val_ptr = (uint16_t*)const_cast<impl_type*>(&val);
+    const impl_type new_val =
+        __float2half(__half2float(const_cast<impl_type&>(val)) +
+                     __half2float(const_cast<const impl_type&>(src.val)));
+    Kokkos::atomic_store(val_ptr, *((uint16_t*)&new_val));
+#endif
+  }
+
+  KOKKOS_FUNCTION
+  void atomic_mul(const volatile half_t src) const volatile {
+#ifdef __CUDA_ARCH__
+#if (CUDA_VERSION < 10000)
+    alignas(4) float lhs_tmp = __half2float(val),
+                     rhs_tmp = __half2float(src.val);
+    volatile float* lhs      = &lhs_tmp;
+    volatile float* rhs      = &rhs_tmp;
+    val                      = __float2half(lhs[0] * rhs[0]);
+#else
+    // static volatile unsigned mutex = 0; // TODO: this should not be a class
+    // member. while (mutex == 1) ; mutex = 1; val = const_cast<impl_type&>(val)
+    // * const_cast<const impl_type&>(src.val); mutex = 0;
+
+    // TODO: this will not work when alignas(4) is removed.
+    // volatile float* lhs = (float*)&val;
+    // volatile float* rhs = (float*)&src.val;
+    // *lhs = __half2float(val);
+    // *rhs = __half2float(src.val);
+    // lhs[0] *= rhs[0];
+    // val = __float2half(*lhs);
+
+    // 16 bit arithmetic on device follow by 16bit atomic store.
+    uint16_t* val_ptr = (uint16_t*)const_cast<impl_type*>(&val);
+    const impl_type new_val =
+        const_cast<impl_type&>(val) * const_cast<const impl_type&>(src.val);
+    Kokkos::atomic_store(val_ptr, *((uint16_t*)&new_val));
```

Comment:
maybe clal explicitly the CUDA atomic store?

Reply:
can you point me to the CUDA atomic store in the programming guide?